### PR TITLE
Do not ignore fatal error in AsmParser for inline assembly

### PR DESCRIPTION
**Pull request by axic**
Part of #5992.

### CODE REVIEW

**Comment by axic** [comment]
Should I add a changelog entry?

File: libsolidity/parsing/Parser.cpp [hunk 1]
```diff
@@ -1043,6 +1043,10 @@ ASTPointer<InlineAssembly> Parser::parseInlineAssembly(ASTPointer<ASTString> con
 
 	yul::Parser asmParser(m_errorReporter, yul::EVMDialect::looseAssemblyForEVM());
 	shared_ptr<yul::Block> block = asmParser.parse(m_scanner, true);
+	if (block == nullptr)
+		// TODO: investigate if this can be done in a more cohesive manner.
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: I think this is exactly the right way.

**Comment by chriseth** [comment]
I don't think a changelog entry is necessary here.

File: libsolidity/parsing/Parser.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/parsing/Parser.cpp]: The part confusing to me is this in `AsmParser`:
```
        catch (FatalError const&)
        {
                if (m_errorReporter.errors().empty())
                        throw; // Something is weird here, rather throw again.
        }
```

I think what this should be is an assertion instead and then it is consistent. `AsmParser` will not throw a `FatalError` exception ever, just a `nullptr` if unparseable - but may have assertions.

How do you feel changing and removing the comment in the solidity parser?

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: we can change it to an assertion, yes.

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: Which comment?

Comment (axic) [reply libsolidity/parsing/Parser.cpp]: This one you commented to. Did change.

**Comment by axic** [comment]
@chriseth ready for review

**Review by chriseth — APPROVED** [review]

**Pull request merged**